vi.py

```diff
@@ -0,0 +1,224 @@
+import torch
+import dgl
+
+class StagVIRNodeClassifiction(torch.nn.Module):
+    def __init__(
+            self,
+            layer,
+            in_features,
+            hidden_features,
+            out_features,
+            depth,
+            activation=None,
+            a_prior=1.0,
+            a_log_sigma_init=1.0,
+            kl_scaling=1.0,
+        ):
+        super(StagVIRNodeClassifiction, self).__init__()
+
+        # get activation function from pytorch if specified
+        if activation is not None:
+            activation = getattr(torch.nn.functional, activation)
+
+        # initial layer: in -> hidden
+        self.gn0 = layer(
+            in_feats=in_features,
+            out_feats=hidden_features,
+            activation=activation,
+        )
+
+        self.bn0 = torch.nn.BatchNorm1d(hidden_features)
+
+        # last layer: hidden -> out
+        setattr(
+            self,
+            "gn%s" % (depth-1),
+            layer(
+                in_feats=hidden_features,
+                out_feats=hidden_features,
+                activation=None,
+            )
+        )
+
+        # middle layers: hidden -> hidden
+        for idx in range(1, depth-1):
+            setattr(
+                self,
+                "gn%s" % idx,
+                layer(
+                    in_feats=hidden_features,
+                    out_feats=hidden_features,
+                    activation=activation
+                )
+            )
+
+            setattr(
+                self,
+                "bn%s" % idx,
+                torch.nn.BatchNorm1d(hidden_features),
+            )
+
+        if depth == 1:
+            self.gn0 = layer(
+                in_feats=in_features, out_feats=out_features, activation=None
+            )
+
+        self.a_prior = torch.distributions.Normal(1.0, a_prior)
+
+        # middle layers
+        self.a_mu = torch.nn.Parameter(torch.tensor(1.0))
+
+        self.a_log_sigma = torch.nn.Parameter(
+            torch.tensor(a_log_sigma_init)
+        )
+
+        self.in_features = in_features
+        self.hidden_features = hidden_features
+        self.depth = depth
+        self.kl_scaling = kl_scaling
+
+
+
+    def _forward(self, g, x):
+        """ Internal forward.
+
+        Parameters
+        ----------
+        g : dgl.DGLGraph
+            Input graph.
+
+        x : torch.Tensor
+            Input features. (n_nodes, in_features)
+
+        Returns
+        -------
+        dgl.DGLGraph
+            Graph with perturbed weights annotated.
+
+        """
+        g = g.local_var()
```

Comment:
please do sth like g_ = g.local_var() to ensure that we do not ovewrwrite anything outside, OR show me this is harmless with a little toy example.